### PR TITLE
8160404: RelocationHolder constructors have bugs

### PR DESCRIPTION
**Pull request by kimbarrett**
Please review this change to construction and copying of the Relocation and
RelocationHolder classes, to eliminate some questionable C++ usage.

The array type for RelocationHandle::_relocbuf is changed from void* to char,
because using a char array for raw memory is countenanced by the standard,
while not so much for an array of void*.  The desired alignment is maintained
via a union, since using alignas is not (yet) permitted in HotSpot code.

There is also now a comment discussing the use of _relocbuf in more detail,
including some areas of continued sketchiness wrto standard conformance and
reliance on implementation dependent behavior. 

No longer use trivial copy and assignment for RelocationHolder, since that
isn't technically correct.  The Relocation in the holder is not trivially
copyable, since it is polymorphic.  It seemed to work in practice with the
supported compilers, but we shouldn't (and don't need to) rely on it.  Instead
we have a new virtual function Relocation::copy_into that copies the most
derived object into the holder's _relocbuf using placement new.

Eliminated the implict conversion constructor from Relocation to holder that
wordwise copied (to possibly beyond the end of) the Relocation into the
holder's _relocbuf.  We could have implemented this more carefully with the
new approach (using copy_into), but we don't actually need this conversion.
The only use of it was actually a wasted copy (in assembler_x86.cpp).

Eliminated the use of placement new syntax via operator new with a holder
argument to copy a Resource object into a holder.  This included runtime
verification that the size of the object is not greater than the size of
_relocbuf; we now do corresponding verification at compile-time.  This also
included an incorrect attempt at a runtime check that the Relocation base
class would be at the same address as the derived class being constructed; we
now perform that check correctly.  We also discuss in a comment the layout
assumption being made (that isn't required by the standard but is provided by
all supported compilers), and what to do if we encounter a compiler that
behaves differently.

Eliminated the idiom of making a default-constructed holder and then
overwriting its held relocation with a newly constructed one, using the afore
mentioned (and eliminated) operator new.  Instead, RelocationHolder now has a
factory function template (construct<T>) for creating holders with a given
relocation type, constructed using provided arguments.  (The arguments are taken
as const-ref rather than using perfect forwarding, as the tools for the latter
are not (yet) approved for use in HotSpot. Using const-ref is good enough in
this case.)

Describe and verify other assumptions being made, such as all Relocation
classes being trivially destructible.

Testing:
mach5 tier1-5

Future work:

* RelocationHolder::reloc isn't const-correct.  Making it so will require
adjustment of some callers.  I'll follow up with an RFE to address this.

* Relocation classes have many virtual function overrides that are unmarked.
I'll follow up with an RFE to add "override" specifiers.

Potential issue: The removal of RelocationHolder(Relocation*) might not work
for some platforms.  I've tested on platforms supported by Oracle (where there
was only one (mistaken) use).  There might be uses by other platforms.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8160404](https://bugs.openjdk.org/browse/JDK-8160404): RelocationHolder constructors have bugs

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [254b094a](https://git.openjdk.org/jdk/pull/11618/files/254b094a8a1fe741d0d70bb2b59d8abbca736c85)
 * [John R Rose](https://openjdk.org/census#jrose) (@rose00 - **Reviewer**) ⚠️ Review applies to [ebe6e01d](https://git.openjdk.org/jdk/pull/11618/files/ebe6e01d2c632224d200ff8c5ed421504e43a1e3)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) ⚠️ Review applies to [f9c0642e](https://git.openjdk.org/jdk/pull/11618/files/f9c0642e949637ef0281ba93b12fc4f2a7cc0ab6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11618/head:pull/11618` \
`$ git checkout pull/11618`

Update a local copy of the PR: \
`$ git checkout pull/11618` \
`$ git pull https://git.openjdk.org/jdk pull/11618/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11618`

View PR using the GUI difftool: \
`$ git pr show -t 11618`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11618.diff">https://git.openjdk.org/jdk/pull/11618.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Nice cleanup. Thank you, Kim.

**Review by vnkozlov — COMMENTED** [review]
Did you tested it in mach5?

**Comment by kimbarrett** [comment]
> Did you tested it in mach5?

Yes, tier1-5, as mentioned in the PR description (though somewhat buried in there).

**Review by vnkozlov — APPROVED** [review]
Good.

You need second review.

File: src/hotspot/share/code/relocInfo.cpp
Comment (stefank) [inline src/hotspot/share/code/relocInfo.cpp:39]: Please add an empty blank line between HotSpot includes and system includes. We don't explicitly state this, but this was the style that we generated when IncludeDB was removed.

File: src/hotspot/share/code/relocInfo.hpp
Comment (stefank) [inline src/hotspot/share/code/relocInfo.hpp:33]: Add blank line

**Review by stefank — CHANGES_REQUESTED** [review]
Not a review, but a comment about the style for including system includes.

File: src/hotspot/share/code/relocInfo.cpp
Comment (kimbarrett) [reply src/hotspot/share/code/relocInfo.cpp:40]: Done.

File: src/hotspot/share/code/relocInfo.hpp
Comment (kimbarrett) [reply src/hotspot/share/code/relocInfo.hpp:34]: Done.

File: src/hotspot/share/code/relocInfo.cpp
Comment (stefank) [reply src/hotspot/share/code/relocInfo.cpp:40]: Thanks.

**Comment by rose00** [comment]
I just saw the query from Dean Long in the bug comments.  Here is some background information.

This "new relocations" work, a long time ago, was an attempt to add some object oriented structure to the access to relocations.  At the time, I didn't trust the resource allocation mechanism, but I wanted a way to decode "real C++ objects" from a relocation stream.  (Being "real" means supporting virtual methods and a strongly typed API that includes such methods.)  The requirement was to be able to iterate quickly over a compressed stream of reloc info and expose the relevant parts as temporary objects to the client.  (This is still a good design pattern; nowadays we have the more successful field stream API, and perhaps there is more to come.)  In order to make iteration over reloc info fast, I decided to allow those temporary objects to unpack themselves (from the compressed reloc info data) into a buffer on the stack rather than on the heap.  The theory is that as you iterate over the stream, you are using just one cache line to access each record, and you get the benefit of a "real" object API.  I think sometimes this is called the design pattern of "flyweight objects".

At the time, back in the day, I knew that I was coloring outside the lines of C++, but I knew (or hoped) that C++ compilers were not smart enough to detect my transgression beyond the language.  Those days are gone now.

I am grateful for the current cleanup.  I hope that it can preserve the basic idea of a "flyweight object", which is an object whose storage is inside a fixed cache line associated with a stream, and whose lifetime ends when the stream is advanced, to the next flyweight object in line.  I hope this because I think that compressed streams are here to stay in the JVM, and yet it is desirable to present the contents of such streams as (temporarily unpacked flyweight) C++ objects, real objects.

Here's another peek beyond the horizon, according to me (and this is just me now):  I think that the best representation for JVM metadata is (often, maybe even always) something that is stream-oriented and position independent and pointer-free.  For example, when we boot up a JVM image, if that image has JVM metadata that was "cooked" before startup, we need to quickly warm up that data so it works in the current JVM process.  In such a situation, today's standard C++ data, which is rich in C pointers (which are machine addresses in full machine words that name positions in virtual memory mappings) has a cost; you need to either make sure that the existing pointer values (as seen during CDS generation or Leyden condensation) are still valid, or else you need to edit all such pointer values so they point to the right place in the new address space.  (This is what relocation records do in C++.  Not the same as this PR.)  This means that today's standard C++ data is actually expensive to build ahead of time, before bootstrap. It would be better if the JVM just accessed blocks of data (probably compressed) which use offsets or other non-pointer means to tie themselves together, like today's reloc info or field info streams.

If you buy all that, then there is a permanent place for "flyweight objects" in the JVM, because when you unpack some sort of position-independent data (pointer free data) you probably want to present it to the client as a "real C++ object".  But in many cases, it is OK if that object has a very short lifetime, and correspondingly constrained allocation, to the workflow of the stream abstraction that is stepping through the pointer-free data.

And that is another reason I'm grateful that Kim has read the C++ manual and figured out a way to do the flyweight objects of reloc info within the C++ rules.

File: src/hotspot/share/code/relocInfo.hpp
Comment (JornVernee) [inline src/hotspot/share/code/relocInfo.hpp:859]: Should this be:
```suggestion
  // types is trivial, so can be non-virtual.
```
?

File: src/hotspot/share/code/relocInfo.hpp
Comment (JornVernee) [inline src/hotspot/share/code/relocInfo.hpp:528]: It seems that the current code is like this, rather than using `typename T, typename...Args` (where `T` is the relocation type), because it is not possible to specify explicit type arguments when calling a constructor template? (So you can't specify `T`).

But, you could truck in the type argument on `Construct`, and have it be inferred. So, this could become:

```suggestion
  template<typename T>
  struct Construct {};          // Tag for selecting this constructor.
  template<typename T, typename... Args> RelocationHolder(Construct<T>, const Args&...args) {
    check_reloc_type<T>();
    Relocation* reloc = ::new (_relocbuf) T(args...);
    check_reloc_placement(reloc);
  }
```
Callers provide e.g. `Construct<T>()` in `construct` or `Construct<Relocation>()` in the default constructor. `construct` can just forward the rest of the args.

This seems a little simpler (by avoiding the use of lambdas, and making it more locally obvious that the constructor is doing a placement new into `_relocbuf`), so maybe this is preferable?

File: src/hotspot/share/code/relocInfo.hpp
Comment (JornVernee) [inline src/hotspot/share/code/relocInfo.hpp:544]: This would become
```suggestion
  static RelocationHolder construct(const Args&... args) {
    return RelocationHolder(Construct<T>(), args...);
  }
```

File: src/hotspot/share/code/relocInfo.hpp
Comment (JornVernee) [inline src/hotspot/share/code/relocInfo.hpp:892]: And this would become
```suggestion
inline RelocationHolder::RelocationHolder() : RelocationHolder(Construct<Relocation>())
{}
```

**Comment by rose00** [comment]
Looks good.  Thanks for bringing this code into the future.

I suggest putting all the static asserts in one place, instead of two of them in place A and the third (trivial destructor) in place B.

Task for another day:  Figure out best practices for doing flyweight objects in C++, and package them up for application elsewhere.  I think Valhalla could use virtual flyweight field descriptors.

Alternative task for another day:  Figure out best practices for *avoiding* flyweight objects, and use those instead.

You can mark me as a reviewer.

File: src/hotspot/share/code/relocInfo.hpp
Comment (kimbarrett) [reply src/hotspot/share/code/relocInfo.hpp]: Yes, that works, and it's an improvement. Thanks. It also let me merge most of
the implementation of that constructor with the implementation of
`copy_into_impl`, which now use new `emplace_resource`.

I also took a moment to use `alignas` instead of a union, since JDK-8297912
was integrated earlier today.

**Review by rose00 — APPROVED** [review]

File: src/hotspot/share/code/relocInfo.hpp
Comment (dean-long) [inline src/hotspot/share/code/relocInfo.hpp:1041]: I'd rather see (metadata_index, offset) than (0, 0), but I guess the meaning of the argument can be found in spec() above.

**Comment by kimbarrett** [comment]
> I suggest putting all the static asserts in one place, instead of two of them in place A and the third (trivial destructor) in place B.

The static asserts about class relationships and size are checked where they are relied upon.  The trivial destructor assumption is relied upon elsewhere (holder assignment, copy_into implementations, and ~Relocation) and not there.  There isn't really a single good place to verify assumption.  I've expanded on comments to indicate where that assumption comes into play.

File: src/hotspot/share/code/relocInfo.hpp
Comment (JornVernee) [reply src/hotspot/share/code/relocInfo.hpp]: Ah, ok. Thanks for explaining.

**Review by JornVernee — APPROVED** [review]

**Comment by kimbarrett** [comment]
Thanks all for reviews and comments.

File: src/hotspot/share/code/relocInfo.hpp
Comment (rose00) [reply src/hotspot/share/code/relocInfo.hpp]: Nice suggestion.  That makes the whole pattern a little more palatable, since it puts the magic new in one place instead of N places.

**Pull request closed without merging**
